tokio/src/task/scope/scope.rs

```diff
@@ -0,0 +1,354 @@
+
+                // TODOs:
+                // - This should not have a futures dependency
+                // - This might block multithreaded runtimes, since the tasks might need
+                //   the current executor thread to make progress, due to dependening on
+                //   its IO handles. We need to do something along task::block_in_place
+                //   to solve this.
+                futures::executor::block_on(wait_fut);
+            }
+            ScopeDropBehavior::Panic => {
+                panic!("Scope was dropped before child tasks run to completion");
+            }
+            ScopeDropBehavior::Abort => {
+                eprintln!("[ERROR] A scope was dropped without being awaited");
+                std::process::abort();
+            }
+            ScopeDropBehavior::ContinueTasks => {
+                // Do nothing
+            }
+        }
+    }
+}
+
+impl ScopeHandle {
+    /// spawns a task on the scope
+    pub fn spawn<'inner, T, R>(&'inner self, task: T) -> ScopedJoinHandle<'inner, R>
+    where
+        T: Future<Output = R> + Send + 'static,
+        R: Send + 'static,
+        T: 'inner,
+    {
+        let spawn_handle =
+            crate::runtime::context::spawn_handle().expect("Spawn handle must be available");
+
+        let releaser = self.scope.wait_group.add();
+        let cancel_token = self.scope.cancel_token.clone();
+        let cancel_behavior = self.scope.options.cancel_behavior;
+
+        let child_task = {
+            spawn_handle.spawn(async move {
+                // Drop this at the end of the task to signal we are done and unblock
+                // the WaitGroup
+                let _wait_group_releaser = releaser;
+
+                if cancel_behavior == ScopeCancelBehavior::CancelChildTasks {
+                    futures::pin_mut!(task);
+                    use futures::FutureExt;
+
+                    futures::select! {
+                        _ = cancel_token.wait().fuse() => {
+                            // The child task was cancelled
+                            return Err(CancellableFutureError::Cancelled);
+                        },
+                        res = task.fuse() => {
+                            return Ok(res);
+                        }
+                    }
+                } else {
+                    Ok(task.await)
+                }
+            })
+        };
+
+        // Since `Scope` is `Sync` and `Send` cancellations can happen at any time
+        // in case of invalid use. Therefore we only check cancellations once:
+        // After the task has been spawned. Since the cancellation is already set,
+        // we need to wait for the task to complete. Then we panic due to invalid
+        // API usage.
+        if self.scope.is_cancelled() {
+            futures::executor::block_on(async {
+                let _ = child_task.await;
+            });
+            panic!("Spawn on cancelled Scope");
```

Comment:
Since this panic will signal incorrect API use to a user, it would be nice if the panic message was a little more descriptive.